utils.bash

```diff
@@ -0,0 +1,147 @@
+# Not an executable. Can be sourced by other scripts in this repo for some free
+# error checking for common operations. Uses bash-specific features including
+# `local`.
+
```

Comment:
This is sortof problematic, because the idea of these scripts is to have reproducibility moving forward. If someone were to come along and edit this and break old scripts, we would be fairly hosed trying to rebuild old binaries.

Could we maybe get ahead of that by versioning this? `utils.v1.bash`, and then just committing to (almost) never changing it?

Reply:
In general though, a bit more copypasta is fine here.

Reply:
What if instead of `source`ing whatever's in the current version of `utils.bash`, a script was to request it from git by SHA? e.g.:
```
PAGER=cat git show 306a85117b67e518513e21439e7ddc4eb1fcab9b:utils.bash > utils-306a85117b67e518513e21439e7ddc4eb1fcab9b.bash
source utils-306a85117b67e518513e21439e7ddc4eb1fcab9b.bash
```

This is only a little more work, and resolves the issue of reproducibility while allowing us to maintain and augment the `utils.bash` script. This could even be done in `build.sh` scripts for individual versions of e.g. clang which invoke the `build-clang.sh` script at the root of the repo, as in:
```
PAGER=cat git show 306a85117b67e518513e21439e7ddc4eb1fcab9b:build-clang.sh > build-clang-306a85117b67e518513e21439e7ddc4eb1fcab9b.sh
chmod +x build-clang-306a85117b67e518513e21439e7ddc4eb1fcab9b.sh
./build-clang-306a85117b67e518513e21439e7ddc4eb1fcab9b.sh osx 6.0.0
```

The path to the right of the colon in `git show <sha>:<script-path>` is always relative to the repo root, so this technique would also mean you could run e.g. `build-support/bin/clang/mac/10.13/6.0.0/build.sh` in the same directory if you wanted.

This technique introduces a dependency on git and on not rewriting the history after commits made to master, which are both reasonable enough assumptions to make, I think? This doesn't seem too awkward to use (especially since it's opt-in), it's definitely portable to all supported platforms, and it allows for perfect reproducibility.

Separately, I think at some point (not now!!) we should probably have some sort of testing for each version of the binaries that we publish, which could potentially be modelled after the kind of testing we perform with the IntelliJ plugin, by [cloning and running tests with multiple different versions of Pants](https://github.com/pantsbuild/intellij-pants-plugin/blob/67c6d81fbeba11869713a0338fdc12e04e2ed6f9/.travis.yml#L40). I think this is probably what we want to aim for in terms of reproducibility as well.

Reply:
Ooor just put a version in the name =)